Today is a very important day because we'll finally connect:   

BERT   
↓    
Embeddings   
↓    
Sentence Transformers   
↓    
RAG   
↓   
Vector Databases   

**Part-1 Token Embeddings**    
* When BERT processes:   
I love Machine Learning    
it creates an embedding for each token.    
* Example:   
I           → Vector 1   
love        → Vector 2  
Machine     → Vector 3    
Learning    → Vector 4   
This is called Token Embeddings    
Output shape:   
torch.Size([1, 6, 768])   
Every token gets its own vector.

**Part-2 Why Token Embeddings Are Not Enough**   
* Suppose we have:    
I love AI    
and   
AI is amazing   
If we want to compare these two sentences, comparing token-by-token is difficult.    
* We need:    
One vector   
for one sentence  

**Part-3 Sentence Embeddings**   
Sentence Transformers solve this problem.   
Instead of:   
Token1 → vector   
Token2 → vector   
Token3 → vector   
* they produce:    
Whole Sentence   
↓  
Single Embedding Vector  
* Example:  
"I love AI"   
↓    
384-dimensional vector

**Exercise 1**

In [8]:
from sentence_transformers import SentenceTransformer

In [9]:
model = SentenceTransformer("all-MiniLM-L6-v2")

embedding = model.encode("I love AI")

print(embedding.shape)

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

(384,)


(384,) Only one vector for the entire sentence.

**Exercise 2**

In [11]:
from sklearn.metrics.pairwise import cosine_similarity

s1 = model.encode("I love Artificial Intelligence")
s2 = model.encode("I enjoy AI")
s3 = model.encode("The weather is very hot")

print(cosine_similarity([s1], [s2]))
print(cosine_similarity([s1], [s3]))

[[0.80237925]]
[[0.02306946]]


s1 vs s2 → High similarity    [[0.80237925]]

s1 vs s3 → Low similarity     [[0.02306946]]

This is semantic understanding.

**A little info about rag**   
What is RAG?   
RAG = Retrieval-Augmented Generation   
It's a technique that combines two things:    
Retrieval fetching relevant information from an external knowledge base    
Generation using an LLM (like GPT) to generate a response based on that retrieved info

**Exercise 3**

In [12]:
sentences = [
    "Machine Learning is interesting",
    "Artificial Intelligence is fascinating",
    "I enjoy studying NLP",
    "Today is very hot"
]

#generate embeddings
sent_embedding = model.encode(sentences)

# Compute pairwise similarities
similarity_matrix = cosine_similarity(sent_embedding)

# Print all unique pairs
print("Pairwise Similarities:\n")
for i in range(len(sentences)):
    for j in range(len(sentences)):
        if i < j:
            print(f"{sentences[i]}")
            print(f"{sentences[j]}")
            print(f"Score: {similarity_matrix[i][j]:.4f}\n")

Pairwise Similarities:

Machine Learning is interesting
Artificial Intelligence is fascinating
Score: 0.7356

Machine Learning is interesting
I enjoy studying NLP
Score: 0.3346

Machine Learning is interesting
Today is very hot
Score: 0.1506

Artificial Intelligence is fascinating
I enjoy studying NLP
Score: 0.3209

Artificial Intelligence is fascinating
Today is very hot
Score: 0.1122

I enjoy studying NLP
Today is very hot
Score: 0.1348

